File: android/app/src/main/java/com/buildapp/photos/data/Uploader.kt
package com.buildapp.photos.data

import android.content.Context
import android.net.Uri
import android.provider.MediaStore
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okio.BufferedSink
import okio.source
import java.util.concurrent.TimeUnit

/**
 * Talks to /api/upload and /api/upload/check. One request per file: a lost
 * Wi-Fi packet mid-batch used to void the whole batch (the server only saves
 * once the entire multipart body has arrived), and per-file requests give
 * real progress.
 */
object Uploader {
    data class FileResult(
        val name: String,
        val ok: Boolean,
        val duplicate: Boolean = false,
        val serverId: String? = null,
        val error: String? = null,
    )

    @Serializable
    data class CheckFile(val name: String, val size: Long, val hash: String? = null)

    @Serializable
    private data class CheckRequest(val files: List<CheckFile>)

    @Serializable
    private data class CheckResponse(val ok: Boolean, val exists: List<Boolean>, val ids: List<String?>)

    @Serializable
    private data class UploadItem(
        val id: String? = null,
        val name: String,
        val indexed: Boolean = false,
        val duplicate: Boolean = false,
        val reason: String? = null,
    )

    @Serializable
    private data class UploadResponse(val ok: Boolean, val count: Int, val items: List<UploadItem>)

    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false }

    private val client = OkHttpClient.Builder()
        .addInterceptor(com.buildapp.photos.api.Auth.interceptor)
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.MINUTES)
        .writeTimeout(10, TimeUnit.MINUTES)
        .retryOnConnectionFailure(true)
        .build()

    fun base(serverUrl: String) = if (serverUrl.endsWith("/")) serverUrl else "$serverUrl/"

    /**
     * Which of these files does the server already have? With a content
     * hash the match is by bytes regardless of name; without one the server
     * falls back to name + size.
     */
    fun check(serverUrl: String, files: List<CheckFile>): Result<List<Boolean>> = runCatching {
        if (files.isEmpty()) return@runCatching emptyList()
        val body = json.encodeToString(CheckRequest.serializer(), CheckRequest(files))
            .toRequestBody("application/json".toMediaType())
        val req = Request.Builder().url(base(serverUrl) + "api/upload/check").post(body).build()
        client.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) error("check failed: HTTP ${resp.code}")
            json.decodeFromString(CheckResponse.serializer(), resp.body!!.string()).exists
        }
    }

    /**
     * Upload a single content:// item. Never throws; the result says what
     * happened. `folder` is the phone folder the file lives in (Camera,
     * WhatsApp Images...) and `device` this phone's name: the server files
     * uploads by them, and uses the phone as the camera for EXIF-less files.
     */
    fun uploadOne(
        context: Context, serverUrl: String, uri: Uri, name: String, size: Long, mime: String?,
        folder: String? = null, device: String? = DeviceInfo.name(context), deviceMake: String? = DeviceInfo.make(),
    ): FileResult {
        val src = Source(folder, device, deviceMake)
        val first = uploadOnce(context, serverUrl, uri, name, size, mime, src)
        // MediaStore's SIZE can lag behind the file (a still-being-written
        // camera burst, an edited-in-place image); OkHttp then aborts with a
        // length mismatch. Send again with chunked encoding, which doesn't
        // need to know the length up front.
        val lengthMismatch = !first.ok && size > 0 &&
            (first.error?.contains("expected", ignoreCase = true) == true ||
                first.error?.contains("Content-Length", ignoreCase = true) == true)
        return if (lengthMismatch) uploadOnce(context, serverUrl, uri, name, -1, mime, src) else first
    }

    private data class Source(val folder: String?, val device: String?, val deviceMake: String?)

    private fun uploadOnce(context: Context, serverUrl: String, uri: Uri, name: String, size: Long, mime: String?, src: Source): FileResult {
        val resolver = context.contentResolver
        val body = object : RequestBody() {
            override fun contentType() = (mime ?: "application/octet-stream").toMediaTypeOrNull()
            // Known length => Content-Length instead of chunked encoding, which
            // lets the server reject oversize early and proxies buffer less.
            override fun contentLength(): Long = if (size > 0) size else -1
            override fun writeTo(sink: BufferedSink) {
                val input = resolver.openInputStream(uri) ?: error("cannot open $uri")
                input.use { sink.writeAll(it.source()) }
            }
        }
        val multipart = MultipartBody.Builder().setType(MultipartBody.FORM).apply {
            src.device?.let { addFormDataPart("device", it) }
            src.deviceMake?.let { addFormDataPart("device_make", it) }
            src.folder?.let { addFormDataPart("folder", it) }
            addFormDataPart("file", name, body)
        }.build()
        val req = Request.Builder().url(base(serverUrl) + "api/upload").post(multipart).build()
        return try {
            client.newCall(req).execute().use { resp ->
                if (!resp.isSuccessful) return FileResult(name, ok = false, error = "HTTP ${resp.code}")
                val parsed = json.decodeFromString(UploadResponse.serializer(), resp.body!!.string())
                val item = parsed.items.firstOrNull()
                    ?: return FileResult(name, ok = false, error = "empty response")
                if (!item.indexed) FileResult(name, ok = false, error = item.reason ?: "not indexed")
                else FileResult(item.name, ok = true, duplicate = item.duplicate, serverId = item.id)
            }
        } catch (e: Exception) {
            FileResult(name, ok = false, error = e.message ?: e.javaClass.simpleName)
        }
    }

    /** Resolve display name, size and mime for any content:// URI (photo picker or MediaStore). */
    fun describe(context: Context, uri: Uri): Triple<String, Long, String?> {
        val resolver = context.contentResolver
        var displayName: String? = null
        var size = -1L
        runCatching {
            resolver.query(uri, arrayOf(MediaStore.MediaColumns.DISPLAY_NAME, MediaStore.MediaColumns.SIZE),
                null, null, null)?.use { c ->
                if (c.moveToFirst()) {
                    displayName = c.getString(0)
                    size = c.getLong(1)
                }
            }
        }
        val name = displayName
            ?: uri.lastPathSegment?.substringAfterLast('/')
            ?: "upload-${System.currentTimeMillis()}"
        return Triple(name, size, resolver.getType(uri))
    }

    /**
     * Manual "pick and upload" path. Uploads one file at a time and reports
     * progress after each; returns every per-file result.
     */
    fun uploadAll(
        context: Context,
        serverUrl: String,
        uris: List<Uri>,
        onProgress: (done: Int, total: Int, last: FileResult) -> Unit = { _, _, _ -> },
    ): List<FileResult> {
        val results = mutableListOf<FileResult>()
        uris.forEachIndexed { i, uri ->
            val (name, size, mime) = describe(context, uri)
            val r = uploadOne(context, serverUrl, uri, name, size, mime)
            results += r
            onProgress(i + 1, uris.size, r)
        }
        return results
    }
}

File: ios/Chitra/Data/Uploader.swift
import Foundation
import Photos
import UniformTypeIdentifiers

/**
 Talks to /api/upload and /api/upload/check. One request per file: a lost
 Wi-Fi packet mid-batch would void a whole batch (the server only saves once
 the entire multipart body has arrived), and per-file requests give real
 progress.

 Each body is staged as a file on disk and uploaded with `uploadTask(fromFile:)`,
 so a 4 GB video never lands in memory and Content-Length is always the real
 length — the Android client's stale-size retry has no counterpart here.
 */
enum Uploader {

    struct FileResult {
        let name: String
        let ok: Bool
        var duplicate: Bool = false
        var serverId: String? = nil
        var error: String? = nil
    }

    struct CheckFile: Encodable {
        let name: String
        let size: Int64
        let hash: String?
    }

    private struct CheckRequest: Encodable { let files: [CheckFile] }

    private struct CheckResponse: Decodable {
        let ok: Bool
        let exists: [Bool]
    }

    private struct UploadItem: Decodable {
        let id: String?
        let name: String
        let indexed: Bool
        let duplicate: Bool
        let reason: String?

        enum CodingKeys: String, CodingKey { case id, name, indexed, duplicate, reason }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            id = c.opt(.id)
            name = c.get(.name, "")
            indexed = c.get(.indexed, false)
            duplicate = c.get(.duplicate, false)
            reason = c.opt(.reason)
        }
    }

    private struct UploadResponse: Decodable {
        let ok: Bool
        let items: [UploadItem]

        enum CodingKeys: String, CodingKey { case ok, items }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            ok = c.get(.ok, false)
            items = c.get(.items, [])
        }
    }

    /// Where the phone says a file came from, so the server can file uploads
    /// by device and album and use the phone as the camera for EXIF-less files.
    struct Source {
        var folder: String?
        var device: String?
        var deviceMake: String? = DeviceInfo.make()
    }

    private static let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 600
        cfg.timeoutIntervalForResource = 3600
        cfg.allowsCellularAccess = true
        return URLSession(configuration: cfg)
    }()

    static func base(_ serverUrl: String) -> String { Urls.base(serverUrl) }

    // MARK: - Pre-flight

    /// Which of these files does the server already have? With a content hash
    /// the match is by bytes regardless of name; without one the server falls
    /// back to name + size.
    static func check(serverUrl: String, files: [CheckFile]) async -> [Bool]? {
        guard !files.isEmpty else { return [] }
        guard let url = URL(string: base(serverUrl) + "api/upload/check") else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONEncoder().encode(CheckRequest(files: files))
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { return nil }
            return try JSONDecoder().decode(CheckResponse.self, from: data).exists
        } catch {
            return nil
        }
    }

    // MARK: - Upload

    /// Upload one library asset. Never throws; the result says what happened.
    static func upload(asset: PHAsset, name: String, serverUrl: String, source: Source) async -> FileResult {
        let staged: URL
        do {
            staged = try await stageMultipart(name: name, mime: mimeType(for: name), source: source) { sink in
                try await DeviceMedia.readData(for: asset, sink: sink)
            }
        } catch {
            return FileResult(name: name, ok: false, error: error.localizedDescription)
        }
        defer { try? FileManager.default.removeItem(at: staged) }
        return await send(staged, name: name, serverUrl: serverUrl)
    }

    /// Upload a file already on disk (the manual picker's fallback path for
    /// items PhotoKit won't hand us as assets).
    static func upload(fileAt fileURL: URL, name: String, serverUrl: String, source: Source) async -> FileResult {
        let staged: URL
        do {
            staged = try await stageMultipart(name: name, mime: mimeType(for: name), source: source) { sink in
                let handle = try FileHandle(forReadingFrom: fileURL)
                defer { try? handle.close() }
                while let chunk = try handle.read(upToCount: 1 << 20), !chunk.isEmpty {
                    sink(chunk)
                }
            }
        } catch {
            return FileResult(name: name, ok: false, error: error.localizedDescription)
        }
        defer { try? FileManager.default.removeItem(at: staged) }
        return await send(staged, name: name, serverUrl: serverUrl)
    }

    private static let boundary = "----ChitraFormBoundary7MA4YWxkTrZu0gW"

    /// Writes the whole multipart body to a temp file, streaming the payload
    /// through `writePayload` so the bytes are never all in memory at once.
    private static func stageMultipart(
        name: String,
        mime: String,
        source: Source,
        writePayload: (@escaping (Data) -> Void) async throws -> Void
    ) async throws -> URL {
        let staged = FileManager.default.temporaryDirectory
            .appendingPathComponent("upload-\(UUID().uuidString).multipart")
        FileManager.default.createFile(atPath: staged.path, contents: nil)
        let handle = try FileHandle(forWritingTo: staged)

        func writeString(_ s: String) { handle.write(Data(s.utf8)) }
        func field(_ key: String, _ value: String) {
            writeString("--\(boundary)\r\nContent-Disposition: form-data; name=\"\(key)\"\r\n\r\n\(value)\r\n")
        }

        if let device = source.device { field("device", device) }
        if let make = source.deviceMake { field("device_make", make) }
        if let folder = source.folder { field("folder", folder) }
        writeString("--\(boundary)\r\nContent-Disposition: form-data; name=\"file\"; filename=\"\(name)\"\r\n")
        writeString("Content-Type: \(mime)\r\n\r\n")

        do {
            try await writePayload { chunk in handle.write(chunk) }
        } catch {
            try? handle.close()
            try? FileManager.default.removeItem(at: staged)
            throw error
        }
        writeString("\r\n--\(boundary)--\r\n")
        try handle.close()
        return staged
    }

    private static func send(_ body: URL, name: String, serverUrl: String) async -> FileResult {
        guard let url = URL(string: base(serverUrl) + "api/upload") else {
            return FileResult(name: name, ok: false, error: "bad server url")
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        do {
            let (data, response) = try await session.upload(for: request, fromFile: body)
            guard let http = response as? HTTPURLResponse else {
                return FileResult(name: name, ok: false, error: "no response")
            }
            guard (200..<300).contains(http.statusCode) else {
                return FileResult(name: name, ok: false, error: "HTTP \(http.statusCode)")
            }
            let parsed = try JSONDecoder().decode(UploadResponse.self, from: data)
            guard let item = parsed.items.first else {
                return FileResult(name: name, ok: false, error: "empty response")
            }
            guard item.indexed else {
                return FileResult(name: name, ok: false, error: item.reason ?? "not indexed")
            }
            return FileResult(name: item.name.isEmpty ? name : item.name, ok: true,
                              duplicate: item.duplicate, serverId: item.id)
        } catch {
            return FileResult(name: name, ok: false, error: error.localizedDescription)
        }
    }

    static func mimeType(for filename: String) -> String {
        let ext = (filename as NSString).pathExtension
        if let type = UTType(filenameExtension: ext), let mime = type.preferredMIMEType {
            return mime
        }
        return "application/octet-stream"
    }
}
